File: Pset1/Credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main(void)
{
    bool valid = false;
    int cardLength ;
    long num, cardNumber; // num will be a copy of the cardnumber 

    cardLength = 0;
    cardNumber = get_long("Number: ");
    num = cardNumber;

    while (num != 0) // calculation number of digits 
    {
        num /= 10;     
        ++cardLength;
    }
    valid = (cardLength == 15) || (cardLength == 16) || (cardLength == 13); // make sure number of credit is right
    if (valid)
    {
        valid = false;
        int operation [cardLength];
        int i = 0, k = 0;
        num = cardNumber;
        while (num > 0) //seperating cardnumber into individual digits
        {
            int digit = num % 10;
            operation[i++] = digit;
            num /= 10;
        }

        int sum = 0;
        //int evens[cardLength / 2];
        for (i = 0; i < cardLength; i++)
        {
            if (i % 2 == 0)
            {
                sum += operation[i];
                //printf("%i\n", sum);
            }
            else 
            {
                int mux = operation[i] * 2;
                while (mux > 0)
                {
                    int digit = mux % 10;
                    sum += digit;
                    mux /= 10;
                }
                //printf("%i\n", sum);
            }
        }
        /*for (i = 0; i < cardLength/2 - 1; i++)
        {
            while (evens[i] > 0)
            {
                int digit = evens[i] % 10;
                sum += digit;
                evens[i] /= 10;
            }
        }*/
        bool continuing = sum % 10 == 0 ? true : false;
        //printf("%d" , valid);

        if (continuing)
        {
            if (cardLength == 13)      //(cardLength == 16 && ((cardNumber / (long)pow(10, 15)) == 4)))
            {
                if ((cardNumber / (long)pow(10, 12)) == 4)
                {
                    printf("VISA\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
            else if (cardLength == 15)
            {
                int last_two = cardNumber / (long)pow(10, 13);
                if (last_two == 34 || last_two == 37)
                {
                    printf("AMEX\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
            else if (cardLength == 16)
            {
                int last_two = cardNumber / (long)pow(10, 14);
                if (last_two == 51 || last_two == 52 || last_two == 53 || last_two == 54 || last_two == 55)
                {
                    printf("MASTERCARD\n");
                } 
                else if ((cardNumber / (long)pow(10, 15)) == 4)
                {
                    printf("VISA\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: Pset6/credit py/credit.py
from cs50 import get_string


def main():
    card = get_string("Number: ")
    check = check_format(card)
    if (check):
        is_valid = valid(card)
        if (is_valid):
            card_type(card)
        else:
            print("INVALID\n")
    else:
        print("INVALID\n")


def check_format(s):
    if len(s) != 13 and len(s) != 15 and len(s) != 16:
        return False
    else:
        return s.isdigit()


def valid(s):
    sum = 0
    N = len(s)
    s = int(s)
    for i in range(N):
        digit = s % 10
        if i % 2 == 0:
            sum += digit
            # print(sum)
        else:
            digit *= 2
            sum += (digit // 10 + digit % 10)
            # print(sum)
        s //= 10
    if sum % 10 == 0:
        return True
    else:
        return False
        
        
def card_type(s):
    first2 = s[0: 2]
    if len(s) == 13 and s.index(0) == "4":
        print("VISA\n")
    elif len(s) == 15 and (first2 == "37" or first2 == "37"):
        print("AMEX\n")
    elif len(s) == 16:
        if s[0] == "4":
            print("VISA\n")
        elif first2 in ["51", "52", "53", "54", "55"]:
            print("MASTERCARD\n")
    else:
        print("INVALID\n")


main()